### ISSUE TITLE
panic at github.com/rclone/rclone/fs/metadata.go:112: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by chscott** [rclone/rclone#7845]
#### The associated forum post URL from `https://forum.rclone.org`

N/A

#### What is the problem you are having with rclone?

Repeatable crash like this:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x28 pc=0x7c497a]

goroutine 32662 [running]:
github.com/rclone/rclone/fs.metadataMapper({0x2a69050, 0xc0003203c0}, {0xc0000ca000, 0x14, 0x14}, {0x2a7e870, 0xc0017d9e00}, {0x4bd69a08, 0xc0005517a0}, 0xc0016e9c80)
	github.com/rclone/rclone/fs/metadata.go:112 +0x31a
github.com/rclone/rclone/fs.GetMetadataOptions({0x2a69050, 0xc0003203c0}, {0x2a7e870, 0xc0017d9e00}, {0x4bd69a08, 0xc0005517a0}, {0x0, 0x0, 0x4bd33708?})
	github.com/rclone/rclone/fs/metadata.go:166 +0x1e8
github.com/rclone/rclone/backend/local.(*Object).Update(0xc000a5b320, {0x2a69050, 0xc0003203c0}, {0x4bd69988, 0xc0019102e0}, {0x2a746c0, 0xc0005517a0}, {0x0, 0x0, 0x0})
	github.com/rclone/rclone/backend/local/local.go:1423 +0x8d0
github.com/rclone/rclone/backend/local.(*Fs).Put(0xc0017d9e00, {0x2a69050, 0xc0003203c0}, {0x4bd69988, 0xc0019102e0}, {0x2a746c0, 0xc0005517a0}, {0x0, 0x0, 0x0})
	github.com/rclone/rclone/backend/local/local.go:696 +0xbd
github.com/rclone/rclone/backend/local.(*Fs).PutStream(0x40fabf?, {0x2a69050?, 0xc0003203c0?}, {0x4bd69988?, 0xc0019102e0?}, {0x2a746c0?, 0xc0005517a0?}, {0x0, 0x0, 0x0})
	github.com/rclone/rclone/backend/local/local.go:705 +0x65
github.com/rclone/rclone/fs/operations.Rcat({0x2a69050, 0xc0003203c0}, {0x2a7ecd0, 0xc000340840}, {0xc0001211d0, 0x47}, {0x2a5a670, 0xc001323420}, {0x121eac0, 0xedb86ca95, ...}, ...)
	github.com/rclone/rclone/fs/operations/operations.go:1359 +0x1003
github.com/rclone/rclone/fs/operations.(*copy).rcat(0xc0011ca2d0, {0x2a69050, 0xc0003203c0}, {0x2a5a670, 0xc001323420})
	github.com/rclone/rclone/fs/operations/copy.go:183 +0xe9
github.com/rclone/rclone/fs/operations.(*copy).manualCopy(0xc0011ca2d0, {0x2a69050, 0xc0003203c0})
	github.com/rclone/rclone/fs/operations/copy.go:260 +0x616
github.com/rclone/rclone/fs/operations.(*copy).copy(0xc0011ca2d0, {0x2a69050, 0xc0003203c0})
	github.com/rclone/rclone/fs/operations/copy.go:302 +0x16c
github.com/rclone/rclone/fs/operations.Copy({0x2a69050, 0xc0003203c0}, {0x2a7ecd0, 0xc000340840}, {0x0, 0x0}, {0xc0001211d0, 0x47}, {0x2a7da70, 0xc000bb90e0})
	github.com/rclone/rclone/fs/operations/copy.go:404 +0x485
github.com/rclone/rclone/fs/sync.(*syncCopyMove).pairCopyOrMove(0xc000a60288, {0x2a69050, 0xc0003203c0}, 0xc000320320, {0x2a7ecd0, 0xc000340840}, 0x25, 0xc000d2bf60?)
	github.com/rclone/rclone/fs/sync/sync.go:498 +0x205
created by github.com/rclone/rclone/fs/sync.(*syncCopyMove).startTransfers in goroutine 1
	github.com/rclone/rclone/fs/sync/sync.go:528 +0x6b
```

#### What is your rclone version (output from `rclone version`)

`v1.67.0-beta.7919.df0229245.pr-7789-onedrive`

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Windows 2022, 64 bit

#### Which cloud storage system are you using? (e.g. Google Drive)

Google Workspace (source)
OneDrive/SharePoint (target)

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

`rclone copy Source: Target:`

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

I'll email logs to the Support address shortly.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
Rclone crashed here

https://github.com/rclone/rclone/blob/6a0a54ab97a5aabadf0444497629a8254dc4df3f/fs/metadata.go#L112

Which means most likely that `fInfo` is `nil`, which is weird. This was created a few lines earlier here

https://github.com/rclone/rclone/blob/6a0a54ab97a5aabadf0444497629a8254dc4df3f/fs/metadata.go#L107

So it looks like the directory entry returned `nil` here.

I'm having trouble working out how this could happen!

It looks like it is triggered by the --metadata-mapper code when writing info about the source. Was the source google drive or onedrive - I am not sure?

I suspect it is caused by google drive and probably one of the special files like a google doc, but I'm not sure. Are there any non-files in the google drive source?

Thanks

**Comment by chscott** [comment]
This is a Google Drive (shared drive) to SharePoint (document library) scenario. Is there a way to correlate the failing thread with the activity on that thread leading up to the crash? I don't know if a line like `goroutine 32662` means a thread with ID `32662`, but even if it does, I don't see that ID mentioned elsewhere in the log, so I'm not sure which file is being processed.

I see these file types in the `Metadata mapper sent` log records:

```
video/quicktime
application/pdf
application/vnd.google-apps.document
application/vnd.openxmlformats-officedocument.spreadsheetml.sheet
application/vnd.openxmlformats-officedocument.wordprocessingml.document
```

For the Google docs, `SrcFs` is `memory:`, and `SrcFsType` is `object.memoryFs`. For example:

```
{
    "SrcFs": "Source{xrgXa}:",
    "SrcFsType": "drive",
    "DstFs": "Target{qpiyZ}:",
    "DstFsType": "onedrive",
    "Remote": "REDACTED.pdf",
    "Size": 194032414,
    "MimeType": "application/pdf",
    "ModTime": "2023-07-26T11:11:00Z",
    "IsDir": false,
    "ID": "REDACTED",
    "Metadata": {
        "btime": "2023-08-04T15:10:10.334Z",
        "content-type": "application/pdf",
        "copy-requires-writer-permission": "false",
        "mtime": "2023-07-26T11:11:00.000Z",
        "starred": "false",
        "viewed-by-me": "true",
        "writers-can-share": "false"
    }
}

{
    "SrcFs": "memory:",
    "SrcFsType": "object.memoryFs",
    "DstFs": "Target{qpiyZ}:",
    "DstFsType": "onedrive",
    "Remote": "REDACTED.docx",
    "Size": 10606,
    "MimeType": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
    "ModTime": "2023-12-01T13:25:54.84Z",
    "IsDir": false,
    "Metadata": {
        "btime": "2023-12-01T13:15:46.499Z",
        "content-type": "application/vnd.google-apps.document",
        "copy-requires-writer-permission": "false",
        "mtime": "2023-12-01T13:25:54.840Z",
        "starred": "false",
        "viewed-by-me": "false",
        "writers-can-share": "false"
    }
}
```

**Comment by ncw** [comment]
> For the Google docs, `SrcFs` is `memory:`, and `SrcFsType` is `object.memoryF

That's odd and probably a clue. Will investigate some more and see if I can reproduce it.

If you set --transfers and --checkers to 1 and that will serialise things and hopefully you'll get to see which file or directory is the problem in the debug. If nothing jumps out, I'll make you a build with more debug in it.

**Comment by ncw** [comment]
I managed to replicate this with

    rclone copy -vvM --streaming-upload-cutoff 0 --metadata-mapper "tee -a /tmp/map.json" drive:GDocs /tmp/drive2 --checkers 1 --transfers 1 --check-first

So it is something to do with Google docs which are larger than the default 100k which is the threshold that we switch from reading into memory to streaming directly.

**Comment by ncw** [comment]
I'll fix this in a bit, but if you are desperate for a workaround use `--streaming-upload-cutoff 10M` where 10M is bigger than the biggest google doc you have.

**Comment by ncw** [comment]
This should fix it - please give it a go!

[v1.67.0-beta.7961.3db983c4d.fix-7845-metadata-panic](https://beta.rclone.org/branch/fix-7845-metadata-panic/v1.67.0-beta.7961.3db983c4d.fix-7845-metadata-panic/) on branch [fix-7845-metadata-panic](https://github.com/rclone/rclone/tree/fix-7845-metadata-panic)

**Comment by chscott** [comment]
I'm not sure if [this comment](https://github.com/rclone/rclone/issues/7848#issuecomment-2113464940) in #7848 means anything for this issue, but the panic no longer occurs, so I'll close on the assumption this specific problem is fixed.

**Comment by ncw** [comment]
Thank you for testing.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.67

#7848 is certainly related but we can fix stuff there.

PS Can you let me close the issue - the open issue reminds me that I need to merge the code! Thanks
